---
title: "abeta_gray_2019"
author: "VGray"
date: "10/20/2019"
output: html_document
---

Load necessary libraries

```{r}

library('dplyr')
library('tidyr')
library('stringr')
library('ggplot2')
library('ggpubr')
library('reshape2')
library('pheatmap')
library('circlize')
library('RColorBrewer')

if (!require("devtools")) {
  install.packages("devtools", dependencies = TRUE)
  library(devtools)
}
install_github("raivokolde/pheatmap")

substr <- function (x, start, stop) 
{
    if (!is.character(x)) 
        x <- as.character(x)
    .Internal(substr(x, as.integer(start), as.integer(stop)))
}

### Set the working directory to your abeta analysis directory
setwd("[/add/your/working/directory/here/]")
```

Figure 2A: Compute replicate correlation

```{r}
## Load dataset with abeta replicate information and deal with the header
abeta_replicates <- read.delim('main_synonymous_scores_shared_full.tsv', sep = '\t', header = F,stringsAsFactors = F)

## Here's what the header looks like
abeta_replicates[1:3,]

## Exclude header from dataset to make our own column names. Set rownames.
abeta_replicates1 <- abeta_replicates[4:dim(abeta_replicates)[1],]
colnames(abeta_replicates1) <- c('mutation','rep1_se','rep1_score','rep2_se','rep2_score','rep3_se','rep3_score')
rownames(abeta_replicates1) <- abeta_replicates1$mutation

## Select single amino acid mutations and exclude WT and synonmyous rows. 
abeta_replicates1.sing <- abeta_replicates1[-grep(',',rownames(abeta_replicates1)),]
abeta_replicates1.sing <- abeta_replicates1.sing[3:dim(abeta_replicates1.sing)[1],]  ## remove wt and syn rows

## Compute replicate correlation & make scatter plots
abeta_replicates1.sing.m <- as.matrix(sapply(abeta_replicates1.sing[,c(3,5,7)], as.numeric))
rownames(abeta_replicates1.sing.m) <- rownames(abeta_replicates1.sing)
abeta_cor <- cor(abeta_replicates1.sing.m, use = 'pairwise.complete')

a <- ggplot(as.data.frame(abeta_replicates1.sing.m), aes(x= rep1_score, y = rep2_score, alpha = .3)) + geom_point() + theme_minimal() + xlim(c(-5,2)) + ylim(c(-5,2)) +  xlab('Replicate 1') + ylab('Replicate 2') + annotate("text", label = paste("R = ",abeta_cor[1,2]), x=-2,y=2,size=3)+ scale_alpha(guide = 'none')
b <- ggplot(as.data.frame(abeta_replicates1.sing.m), aes(x= rep1_score, y = rep3_score, alpha = .3)) + geom_point() + theme_minimal() + xlim(c(-5,2)) + ylim(c(-5,2))+  xlab('Replicate 1') + ylab('Replicate 3')+ annotate("text", label=paste("R = ",abeta_cor[1,3]), x=-2,y=2,size=3)+ scale_alpha(guide = 'none')
c <- ggplot(as.data.frame(abeta_replicates1.sing.m), aes(x= rep2_score, y = rep3_score, alpha = .3)) + geom_point() + theme_minimal() + xlim(c(-5,2)) + ylim(c(-5,2))+  xlab('Replicate 2') + ylab('Replicate 3')+ annotate("text", label=paste("R = ",abeta_cor[2,3]), x=-2,y=2,size=3)+ scale_alpha(guide = 'none')

ggarrange(a,b,c, ncol=3)
dev.off()
```

Figure 2B: Distribution of nonsynonymous and synonymous

```{r}

## Load dataset with abeta replicate information and deal with the header
mtx_all <- read.delim('main_variants_scores.tsv', sep = '\t', header = F,stringsAsFactors = F)

## Here;s what the header looks like. Fix/Format columns and rownames. 
mtx_all[1:2,]
mtx_all1 <- mtx_all[4:dim(mtx_all)[1],]
colnames(mtx_all1) <- c('mutation','se','epsilon', 'score')
rownames(mtx_all1) <- mtx_all1$mutation

### Remove variants with more than one mutation. Identify the synonymous vs. single nucleotide nonsynonymous variants. Do some formatting. 
mtx_all1.sing <- mtx_all1[-grep(',',rownames(mtx_all1)),]  
mtx_all1.sing$mut_type <- NA
mtx_all1.sing[rownames(mtx_all1.sing) %in% rownames(mtx_all1.sing[grep('(p.=)',rownames(mtx_all1.sing)),]),]$mut_type <- "synonymous"
mtx_all1.sing[rownames(mtx_all1.sing) %in% rownames(mtx_all1.sing[-grep('(p.=)',rownames(mtx_all1.sing)),]),]$mut_type <- "nonsynonymous"
mtx_all1.sing$score <- as.numeric(as.character(mtx_all1.sing$score))

### Figure 2B
ggplot(mtx_all1.sing, aes(x = score,  fill = mut_type)) + geom_density(aes(y = ..scaled..)) + theme_classic() + scale_fill_manual( values = c("#2166ac","#b2182b")) +ylab("Density") +xlab("Variant effect score")  + geom_vline(xintercept = 2*sd(mtx_all1.sing[mtx_all1.sing$mut_type == 'synonymous',]$score) +mean(mtx_all1.sing[mtx_all1.sing$mut_type == 'synonymous',]$score), color = 'white')+ geom_vline(xintercept = -2*sd(mtx_all1.sing[mtx_all1.sing$mut_type == 'synonymous',]$score) +mean(mtx_all1.sing[mtx_all1.sing$mut_type == 'synonymous',]$score), color = 'white')+ theme(legend.position="top")

### Define More-aggregating, less-aggregating, wt-like for supplementary Figure 2D
synmin <- -2*sd(mtx_all1.sing[mtx_all1.sing$mut_type == 'synonymous',]$score) + mean(mtx_all1.sing[mtx_all1.sing$mut_type == 'synonymous',]$score)
synmax <- 2*sd(mtx_all1.sing[mtx_all1.sing$mut_type == 'synonymous',]$score) + mean(mtx_all1.sing[mtx_all1.sing$mut_type == 'synonymous',]$score)

mtx_all1.sing$class <- 'wt-like'
mtx_all1.sing[mtx_all1.sing$score < synmin,]$class <- "more aggregating"
mtx_all1.sing[mtx_all1.sing$score > synmax,]$class <- "less aggregating"
positions <- c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43")

```
Supplmentary Figure 2A-B,D 
Figure 2D-F: Abeta solubility score heatmap.

```{r}

## Read in data files & adjust column and row names
abeta <- read.delim('main_synonymous_scores.tsv', sep = '\t', header = F, stringsAsFactors = F)
aa_att <- read.csv('biophysicalAttributes_2014-05-01.csv', header = T)
ss <- read.csv('abeta_dssp_20190119.csv',header = T)
ss$Position <- as.character(ss$Position)

## fix headers, row names, select single amino acid variants
abeta[1:2,]
abeta1 <- abeta[5:dim(abeta)[1],]
colnames(abeta1) <- c('mutation','se','epsilon', 'score')
rownames(abeta1) <- abeta1$mutation
abeta1.sing <- abeta1[-grep(',',rownames(abeta1)),]  ## 817 variants

## Add wt, pos and mt columns to the data frame. Format numeric columns. 
wt <- substring(abeta1.sing$mutation,3,5)
pos<-str_sub(abeta1.sing$mutation,6,-4)
mt <- str_sub(abeta1.sing$mutation,-3,-1)
abeta1.sing1 <- cbind(wt,pos,mt,abeta1.sing)
abeta1.sing1[,c(5:7)] <- sapply(abeta1.sing1[,c(5:7)], as.numeric)

## Reshape the dataframe
abeta1.sing1.mat <- dcast(abeta1.sing1, mt~pos, value.var = 'score')
aminoacids <- abeta1.sing1.mat$mt
abeta1.sing1.mat<- sapply(abeta1.sing1.mat[2:44], as.numeric)
rownames(abeta1.sing1.mat) <- aminoacids
abeta1.sing1.mat.t <- t(abeta1.sing1.mat)

###  Take mean of all, polar and nonpolar amino acid changes
polar <- aa_att[which(aa_att$Hydrophobicity == 'Polar'),]
hydrophobic <- aa_att[which(aa_att$Hydrophobicity == 'Hydrophobic'),]
abeta1.sing2 <- abeta1.sing1[,c(2:3,5:7)]
abeta1.sing2$score <- as.numeric(abeta1.sing2$score)
abeta1.sing2_polar <- abeta1.sing2[abeta1.sing2$mt %in% droplevels(polar$Amino.acid),]
abeta1.sing2_hydro <- abeta1.sing2[abeta1.sing2$mt %in% droplevels(hydrophobic$Amino.acid),]
annotation <- aggregate(abeta1.sing2, by = list(abeta1.sing2$pos), FUN = mean, na.rm = T)
annotation_polar <- aggregate(abeta1.sing2_polar, by = list(abeta1.sing2_polar$pos), FUN = mean, na.rm = T)
annotation_hydro <- aggregate(abeta1.sing2_hydro, by = list(abeta1.sing2_hydro$pos), FUN = mean, na.rm = T)
annotation <- merge(annotation, abeta1.sing1[c('wt','pos')], by.x = 'Group.1', by.y = 'pos')
annotation1 <- merge(annotation, annotation_polar[c('score', 'Group.1')], by = "Group.1")
annotation2 <- merge(annotation1, annotation_hydro[c('score', 'Group.1')], by = "Group.1")
colnames(annotation2) <- c("Group.1","pos","mt","se","epsilon","mean_score","wt","mean_score_polar","mean_score_hydro")
ann1 = annotation2[!duplicated(annotation2$Group.1),]
ann2 <- merge(ann1, aa_att, by.x = 'wt', by.y = 'Amino.acid')
ann2 <- merge(ann2, ss, by.x = 'Group.1', by.y = 'Position', all.x = T)
rownames(ann2) <- ann2$Group.1

### Fix position numbers from 1-43 to 1-42. 
ann2 <- ann2[c("2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43"),]
rownames(ann2) <- c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42")
ann2$Group.1 <- c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42")


### Data formmating and define heatmap parameters. 
dat <- abeta1.sing1.mat[1:21,2:43]

paletteLength <- 50
myColor <- colorRampPalette(c("#2166ac","#92c5de", "white","#f4a582", "#b2182b"))(paletteLength)
# length(breaks) == length(paletteLength) + 1
# use floor and ceiling to deal with even/odd length pallettelengths
myBreaks <- c(seq(-3, 0, length.out=ceiling(paletteLength/2) + 1), 
              seq(1/paletteLength, 1.5, length.out=floor(paletteLength/2)))
#rownames(dat) <- seq(1,42)
dat.r <- dat[,c("2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43")]
colnames(dat.r) <- c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42")
dat.r = dat.r[!rownames(dat.r) %in% c("Ter"),]

## Supplementary figure 2A. Boxplot of solubility scores for each amino acid. 
dat.rm <- melt(dat.r)
ggplot(dat.rm, aes(Var1, value))+geom_boxplot()+geom_jitter(width = 0.15)+theme_classic() +xlab('Mutant amino acid') + ylab("Variant solubility score")

## Figure 2D. 
annotation_colors = list( Hydrophobicity = c(Hydrophobic="#ca0020",
                                            PositiveCharge="#f4a582",
                                            Polar="#ffffff",
                                            negativeCharge="#bababa",
                                            Special="#404040"))

rownames(aa_att) <- aa_att$Amino.acid
map.clustered <- pheatmap(t(dat.r), breaks = myBreaks,
                   color = myColor,
                   show_rownames = T,
                   border_color ='black',
                   cluster_rows = F,
                   na_col = '#A9A9A9',
                   annotation_colors = annotation_colors,
                   annotation_row = ann2[c('Hydrophobicity')], 
                   annotation_col = aa_att[c(8,7,6)]
                  )

### Supplmentary Figure 2B
pos.clustered <- pheatmap(t(dat.r), breaks = myBreaks,
                   color = myColor,
                   show_rownames = T,
                   border_color ='black',
                   cluster_rows = T,
                   na_col = '#A9A9A9',
                   annotation_colors = annotation_colors,
                   annotation_row = ann2[c('Hydrophobicity')], 
                   annotation_col = aa_att[c(8,7,6)]
                  )

## Assign positions to thei respective clusters and compute a few summary statistics
cluster2 <- ann2[c(4,24,10,12,40),]
mean(cluster2$mean_score)
mean(cluster2$mean_score_polar)
mean(cluster2$mean_score_hydro)
cluster1 <- ann2[c(19,20,32,34,17,18,35,36,39,31,41),]
mean(cluster1$mean_score)
mean(cluster1$mean_score_polar)
mean(cluster1$mean_score_hydro)
cluster5 <- ann2[c(14,33,28,30,13,21),]
mean(cluster5$mean_score)
mean(cluster5$mean_score_polar)
mean(cluster5$mean_score_hydro)
cluster3 <- ann2[c(37,38),]
mean(cluster3$mean_score)
mean(cluster3$mean_score_polar)
mean(cluster3$mean_score_hydro)
cluster4 <- ann2[c(1,3,7,42,8,9,5,6,25,29,27,2,26),]
mean(cluster4$mean_score)
mean(cluster4$mean_score_polar)
mean(cluster4$mean_score_hydro)
cluster6 <- ann2[c(11,16,22,15,23),]
mean(cluster6$mean_score)
mean(cluster6$mean_score_polar)
mean(cluster6$mean_score_hydro)
### library stats
length(dat.r[!is.na(dat.r)]) #791/798 possible variants
vars <- (dat.r[!is.na(dat.r)])
length(vars[vars>0.3947705]) ## 246 more soluble than WT
length(vars[vars< -0.2621313]) ## 201 more aggregating than WT
length(vars[vars> -0.2621313 & vars < 0.3947705]) ## WT-like == 344

### Figure 2E
myBreaks2 <- c(seq(-2, 0, length.out=ceiling(paletteLength/2) + 1), 
              seq(1/paletteLength, 1.5, length.out=floor(paletteLength/2)))
anno <- pheatmap(ann2[c('mean_score','mean_score_polar','mean_score_hydro')],
         cluster_rows = F,
         cluster_cols = F,
         breaks = myBreaks,
         color = myColor)

anno <- pheatmap(ann2[c('mean_score','mean_score_polar','mean_score_hydro')],
         cluster_rows = T,
         cluster_cols = F,
         breaks = myBreaks,
         color = myColor)


### Supplementary figure 2D information
### Number of WT-like, more aggregating and less-aggregating than WT
length(dat.r[!is.na(dat.r)]) #791/798 possible variants
vars <- (dat.r[!is.na(dat.r)])
length(vars[vars>0.3947705]) ## 246 more soluble than WT
length(vars[vars< -0.2621313]) ## 201 more aggregating than WT
length(vars[vars> -0.2621313 & vars < 0.3947705]) ## WT-like == 344

dat.rm <- melt(dat)
dat.rm$mut_group <- NA
dat.rm[which(dat.rm$value > 0.3947705),]$mut_group <- "more-soluble than WT"
dat.rm[which(dat.rm$value < -0.2621313),]$mut_group <- "less-soluble than WT"
dat.rm[which(dat.rm$value > -0.2621313 & dat.rm$value< 0.3947705),]$mut_group <- "WT-like"

### Summary for Supplmentary Figure 2D
table(dat.rm$Var2,dat.rm$mut_group)

```

Figure 3C + abeta heatmap

```{r}

## Set heatmap parameters
annotation_colors = list( Hydrophobicity = c(Hydrophobic="#ca0020",
                                            PositiveCharge="#f4a582",
                                            Polar="#ffffff",
                                            negativeCharge="#bababa",
                                            Special="#404040"),
                          X2mxu = c(Strand = "#ca0020",
                                    Bend = '#878787',
                                    Nan = "#ffffff"),
                          X2lmn = c(Strand = "#ca0020",
                                    Bend = '#878787',
                                    Nan = "#ffffff"),
                          X2beg = c(Strand = "#ca0020",
                                    Bend = '#878787',
                                    Nan = "#ffffff"),
                          X2lmp = c(Strand = "#ca0020",
                                    Bend = '#878787',
                                    Nan = "#ffffff",
                                    Bridge = '#bababa'),
                          X1iyt = c(Strand = "#ca0020",
                                    Bend = '#878787',
                                    Nan = "#ffffff",
                                    Helix = '#f4a582',
                                    Turn = '#fddbc7'),
                          X5kk3 = c(Strand = "#ca0020",
                                    Bend = '#878787',
                                    Nan = "#ffffff",
                                    Bridge = '#bababa'),
                          X2m4j = c(Strand = "#ca0020",
                                    Bend = '#878787',
                                    Nan = "#ffffff",
                                    Bridge = '#bababa'),
                          X2lnq = c(Strand = "#ca0020",
                                    Bend = '#878787',
                                    Nan = "#ffffff"),
                          X5oqv = c(Strand = "#ca0020",
                                    Bend = '#878787',
                                    Nan = "#ffffff"),
                          X2nao = c(Strand = "#ca0020",
                                    Bend = '#878787',
                                    Nan = "#ffffff",
                                    Bridge = '#bababa',
                                    Turn = '#fddbc7'))

### Heatmap
rownames(aa_att) <- aa_att$Amino.acid
map.clustered <- pheatmap(t(dat.r), breaks = myBreaks,
                   color = myColor,
                   show_rownames = T,
                   border_color ='black',
                   cluster_rows = F,
                   na_col = '#A9A9A9',
                   annotation_colors = annotation_colors,
                   annotation_row = ann2[c('Hydrophobicity',"X1iyt","X2nao","X2mxu",
                                           "X2m4j","X2lnq","X5kk3",
                                           "X2beg","X2lmn",
                                           "X2lmp", "X5oqv")], 
                   annotation_col = aa_att[c(8,7,6)]
                  )

```

